activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -303,6 +303,151 @@ def _select!(*fields) # :nodoc:
+    #   Post.with_recursive(:threads, non_recursive_query, recursive_query)
+    #   # WITH RECURSIVE threads AS (
+    #   #   SELECT comments.id, comments.parent_id FROM comments WHERE comments.parent_id IS NULL
+    #   #   UNION ALL
+    #   #   SELECT comments.id, comments.parent_id FROM comments JOIN threads ON comments.parent_id = threads.id
+    #   # ) SELECT * FROM posts
+    #
+    # If you need to define columns with your common table name pass it as a `String`.
+    #
+    #   non_recursive_query = Comment.select(:id, :parent_id, 0).where(parent_id: nil)
+    #   recursive_query = Comment.select(:id, :parent_id, "threads.depth + 1").joins("JOIN threads ON threads.id = comments.parent_id")
+    #   Post.with_recursive("threads(id, parent, depth)", non_recursive_query, recursive_query)
+    #   # WITH RECURSIVE threads(id, parent, depth) AS (
+    #   #   SELECT comments.id, comments.parent_id, 0 FROM comments WHERE comments.parent_id IS NULL
+    #   #   UNION ALL
+    #   #   SELECT comments.id, comments.parent_id, threads.depth + 1 FROM comments JOIN threads ON comments.parent_id = threads.id
+    #   # ) SELECT * FROM posts
+    #
+    # Note that `UNION ALL` is used by default. If you need to use `UNION` you can set `union_all` option to `false`.
+    #
+    #   Post.with_recursive("threads(id, parent, depth)", non_recursive_query, recursive_query, union_all: false)
+    #   # WITH RECURSIVE threads(id, parent, depth) AS (
+    #   #   SELECT comments.id, comments.parent_id, 0 FROM comments WHERE comments.parent_id IS NULL
+    #   #   UNION
+    #   #   SELECT comments.id, comments.parent_id, threads.depth + 1 FROM comments JOIN threads ON comments.parent_id = threads.id
+    #   # ) SELECT * FROM posts
+    #
+    # To add multiple CTEs you can chain `.with` / `.with_recursive` calls
+    #
+    #   top_comment = Comment.order(replies_count: :desc).limit(1)
+    #   non_recursive_relation = Comment.select(:id, :parent_id, 0).joins("JOIN top_comment ON top_comment.id = comments.parent_id")
+    #   recursive_relation = Comment.select(:id, :parent_id, "replies.depth + 1").joins("JOIN replies ON comments.parent_id = replies.id")
+    #   Comment
+    #     .with(:top_comment, top_comment)
+    #     .with_recursive(:replies, non_recursive_relation, recursive_relation)
+    #   # WITH RECURSIVE top_comment AS (
+    #   #   SELECT comments.id FROM comments ORDER BY comments.replies_count DESC LIMIT 1
+    #   # ), replies(id, parent_id, depth) AS (
+    #   #   SELECT comments.id, comments.parent_id, 0 FROM comments JOIN top_comment ON top_comment.id = comments.parent_id
+    #   #   UNION ALL
+    #   #   SELECT comments.id, comments.parent_id, replies.depth + 1 FROM comments JOIN replies ON comments.parent_id = replies.id
+    #   # )
+    #   # SELECT comments.* FROM replies
+    #
+    # or pass them as a `Hash` but be aware that `ActiveRecord::Relation` does not have `UNION` support
```

Comment:
I was thinking about introducing UNION support ActiveRecord, this could be great reasoning.